src/compile.rs

```diff
@@ -459,6 +459,197 @@ pub fn validate_rules(statements: &BTreeSet<ValidatedStatement>) -> Result<(), C
     errors.into_result(())
 }
 
+// Go through all functions and check if they are castable
+// based only on their args
+pub fn initialize_castable(functions: &mut FunctionMap, types: &TypeMap) {
+    for func in functions.values_mut() {
+        for arg in &func.args {
+            if arg.param_type.is_associated_resource(types) {
+                func.is_castable = false;
+
+                // If we have found one associated resource
+                // we can continue
+                continue;
+            }
+        }
+    }
+}
+
+// Go through all of the functions and check if they are castable
+// base on functions they call.
+pub fn determine_castable(functions: &mut FunctionMap, types: &TypeMap) -> u64 {
+    let mut num_changed: u64 = 0;
+    // We need tmp_functions to avoid a immutable borrow after a mutable one.
+    let tmp_functions = functions.clone();
+    'outer: for func in functions.values_mut() {
+        // If we are already false there is no reason to check our called functions.
+        if !func.is_castable {
+            continue;
+        }
+        for call in func.original_body {
+            if let Statement::Call(call) = call {
+                if let Some(inner_func) = tmp_functions.get(&call.get_cil_name()) {
+                    if !inner_func.is_castable {
+                        num_changed += 1;
+                        func.is_castable = false;
+                        continue 'outer;
+                    }
+                // The call may be a built-in so check the args
```

Comment:
This comment confuses me.  This is checking the args regardless of whether it's a built-in or not, right?

Reply:
Okay, looking more closely, I think it's not checking the args on non-built-ins, right?  Since this `else` matches the `if` that's a lookup in the function map.  So if we find it in the function map, we just check whether it's marked as castable or not.  If we don't find it in the function map, we check the arguments.

I see in call_casting.cas a line that says `asd<boo>.some_function(foo); // TODO this should break?` which I think may be exactly this case.  From the comment should I infer it's not breaking now?

I think we need to check the args regardless of whether it's a built-in, because otherwise we're not checking what we pass to non-builtins.

Reply:
I made the change to check args all the time, but I do want to talk a bit more about the `asd<boo>.some_function(foo);` call.  I think it should work the way it is written. There are no associated types along that chain.

Reply:
Hmm, I think I may have gotten some_function() and read_boo_tmp_more() mixed up?  I haven't reviewed super thoroughly, but on a quick skim, I think I agree with you now that `asd<boo>.some_function(foo);` looks fine.